In [39]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - Latest 12 weeks of life of the banana plant
---

In [40]:
temperature_l12w = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/Latest_12-weeks_May-4_August-03_2018/' \
                                 'Temperatura_May-4_August-03_2018.csv', )

In [41]:
print(temperature_l12w.shape)
temperature_l12w.head()

(4398, 2)


Fecha:  Temperatura (C)
0  2018-05-04 00:07:38            23.75
1  2018-05-04 00:37:39            23.68
2  2018-05-04 01:07:38            23.68
3  2018-05-04 01:37:39            23.56
4  2018-05-04 02:07:39            23.50

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [42]:
print(temperature_l12w.isnull().any())
temperature_l12w.isnull().values.any()

Fecha:             False
Temperatura (C)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Temperatura (C)` ** and assigning them to the matrix 

`temperature_l12w_array` created such as follow:

In [43]:
temperature_l12w_array = temperature_l12w.iloc[:,1].values

In [44]:
# Rehape the luminosity_luxes array 
temperature_l12w_array = temperature_l12w_array.reshape(-1, 1)
temperature_l12w_array

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

## 1.3 Generating descriptive data to dataset
---

In [45]:
col = ['Temperatura (C)']
temperature_l12w_values_df = pd.DataFrame(temperature_l12w_array, columns=col)
temperature_l12w_values_df_describe = temperature_l12w_values_df.describe()

In [46]:
temperature_l12w_values_df_describe

Temperatura (C)
count      4398.000000
mean         27.531389
std           4.273432
min          21.370000
25%          24.000000
50%          25.900000
75%          31.060000
max          40.250000

In [47]:
# Export this descriptive data to comma separated values and java script object notation
temperature_l12w_values_df_describe.to_csv('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                              'Temperature_Describe_May-4_August-03_2018.cvs', sep=',', header=True, index=True)
temperature_l12w_values_df_describe.to_json('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n'
                                     'Temperature_Describe_May-4_August-03_2018.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`temperature_l12w_values_df`** dataset with 4398 samples rows.

In [48]:
print(temperature_l12w_values_df.shape)
temperature_l12w_values_df.head()

(4398, 1)


Temperatura (C)
0            23.75
1            23.68
2            23.68
3            23.56
4            23.50

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`temperature_l12w_values_df`** dataframe to numpy array such as follow:

In [49]:
# numpy_temperature_values = temperature_values_df.reset_index().values
numpy_temperature_l12w_values = temperature_l12w_values_df.values

In [50]:
numpy_temperature_l12w_values

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

We compose the following datasets from **`numpy_temperature_l12w_values`** array :

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_temperature_l12w_values`** which contain all column features of temperature.

Of this way, we pass all data (4390 samples rows) to from them `temperature_l12w_values_train`

and `temperature_l12w_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **4390 * 0.2 = 878 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **4390 * 0.8 = 3512  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing temperature datasets
---

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

In [51]:
temperature_l12w_values_train, temperature_l12w_values_test = train_test_split(numpy_temperature_l12w_values, test_size = 0.2)

In [52]:
# We have 3518 rows to luminosity_luxes_train
print(type(temperature_l12w_values_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , temperature_l12w_values_train.shape)
print('\n')

# And we have 880 rows to luminosity_luxes_test
print(type(temperature_l12w_values_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , temperature_l12w_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (3518, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (880, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations


---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---

We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Temperature (C)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Temperature training dataset

In [53]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', temperature_l12w_values_train)

Remember our temperature training data 
 [[24.87]
 [23.43]
 [36.62]
 ...
 [30.75]
 [34.87]
 [25.56]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `temperature_l12w_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [54]:
minmax_scale_training = scaler.fit(temperature_l12w_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `temperature_l12w_values_train` data are scaled between **0 to 1**  selected range 

In [55]:
# transform precipitations_train data to maximum and mínimum scale value. 
temperature_l12w_minmax_training = minmax_scale_training.transform(temperature_l12w_values_train)

In [56]:
print ("And, these are our scaled data: " + '\n')
temperature_l12w_minmax_training

And, these are our scaled data: 



array([[0.18538136],
       [0.10911017],
       [0.80773305],
       ...,
       [0.49682203],
       [0.71504237],
       [0.22192797]])

In [57]:
print('Temperature Training dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].min()))

print('Temperature Training dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].max()))

Temperature Training dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Training dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_training` numpy array

- We export this array to comma separated values 


In [58]:
temperature_l12w_train_df = pd.DataFrame(temperature_l12w_minmax_training, columns=col)

In [59]:
temperature_l12w_train_df

Temperatura (C)
0            0.185381
1            0.109110
2            0.807733
3            0.168962
4            0.797669
5            0.162076
6            0.493114
7            0.208686
8            0.642479
9            0.254767
10           0.757945
11           0.251589
12           0.695445
13           0.142479
14           0.145657
15           0.231992
16           0.076271
17           0.605932
18           0.695445
19           0.592691
20           0.334216
21           0.145657
22           0.900424
23           0.466631
24           0.102754
25           0.254767
26           0.076271
27           0.228284
28           0.162076
29           0.360699
...               ...
3488         0.105932
3489         0.820975
3490         0.264831
3491         0.556144
3492         0.384004
3493         0.532839
3494         0.592691
3495         0.179025
3496         0.099576
3497         0.635593
3498         0.135593
3499         0.112818
3500         0.820975
3501         0.046610
3502         0.082627
3503         0.148835
3504         0.648835
3505         0.450212
3506         0.523305
3507         0.188559
3508         0.539725
3509         0.516419
3510         0.225106
3511         0.145657
3512         0.221928
3513         0.158898
3514         0.360699
3515         0.496822
3516         0.715042
3517         0.221928

[3518 rows x 1 columns]

In this way we have the dataset `temperature_l12w_train_df` standardized and training, and export it to a .csv file

In [60]:
temperature_l12w_train_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TRAINING_May-4_August-03_2018.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [61]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', temperature_l12w_values_test)

Remember our temperature testing data 
 [[23.56]
 [23.56]
 [25.87]
 [35.81]
 [30.31]
 [23.25]
 [26.43]
 [31.87]
 [34.5 ]
 [24.31]
 [27.43]
 [23.43]
 [34.93]
 [23.56]
 [34.18]
 [23.  ]
 [31.06]
 [23.06]
 [22.75]
 [30.  ]
 [25.25]
 [23.5 ]
 [27.75]
 [25.93]
 [32.93]
 [24.75]
 [22.25]
 [26.37]
 [24.75]
 [34.5 ]
 [26.25]
 [23.56]
 [22.93]
 [26.31]
 [23.93]
 [24.  ]
 [25.  ]
 [32.06]
 [23.68]
 [24.31]
 [28.62]
 [22.81]
 [23.06]
 [23.62]
 [24.62]
 [22.25]
 [24.37]
 [31.18]
 [24.37]
 [31.75]
 [31.06]
 [22.25]
 [23.25]
 [24.68]
 [27.62]
 [31.56]
 [32.25]
 [35.06]
 [24.18]
 [26.56]
 [23.62]
 [33.06]
 [24.43]
 [28.25]
 [23.93]
 [23.81]
 [25.06]
 [23.31]
 [28.31]
 [31.75]
 [23.93]
 [24.93]
 [32.56]
 [32.5 ]
 [34.18]
 [24.87]
 [25.87]
 [23.87]
 [23.5 ]
 [25.75]
 [24.  ]
 [37.5 ]
 [23.31]
 [24.18]
 [38.56]
 [34.87]
 [29.25]
 [27.25]
 [24.  ]
 [24.  ]
 [23.5 ]
 [28.  ]
 [25.12]
 [27.56]
 [24.5 ]
 [23.18]
 [23.62]
 [33.12]
 [31.18]
 [23.37]
 [35.62]
 [22.5 ]
 [26.25]
 [23.75]
 [32.31]
 [24.37]
 [27. 

In [62]:
minmax_scale_test = scaler.fit(temperature_l12w_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
temperature_l12w_minmax_test = minmax_scale_test.transform(temperature_l12w_values_test)

In [63]:
print ("And, these are our testing scaled data: " + '\n')
temperature_l12w_minmax_test

And, these are our testing scaled data: 



array([[0.10479376],
       [0.10479376],
       [0.2335563 ],
       [0.78762542],
       [0.48104794],
       [0.08751394],
       [0.26477146],
       [0.56800446],
       [0.71460424],
       [0.14659978],
       [0.32051282],
       [0.09754738],
       [0.73857302],
       [0.10479376],
       [0.696767  ],
       [0.0735786 ],
       [0.52285396],
       [0.07692308],
       [0.05964326],
       [0.46376812],
       [0.19899666],
       [0.10144928],
       [0.33835006],
       [0.23690078],
       [0.6270903 ],
       [0.17112598],
       [0.03177258],
       [0.26142698],
       [0.17112598],
       [0.71460424],
       [0.25473802],
       [0.10479376],
       [0.0696767 ],
       [0.2580825 ],
       [0.12541806],
       [0.12931996],
       [0.18506132],
       [0.57859532],
       [0.11148272],
       [0.14659978],
       [0.38684504],
       [0.06298774],
       [0.07692308],
       [0.10813824],
       [0.1638796 ],
       [0.03177258],
       [0.14994426],
       [0.529

In [64]:
print('Temperature Testing dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].min()))

print('Temperature Testing dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].max()))

Temperature Testing dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Testing dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_test` numpy array

- We export this array to comma separated values

In [65]:
temperature_l12w_test_df = pd.DataFrame(temperature_l12w_minmax_test, columns=col)

In [66]:
print(temperature_l12w_test_df.shape)
temperature_l12w_test_df

(880, 1)


Temperatura (C)
0           0.104794
1           0.104794
2           0.233556
3           0.787625
4           0.481048
5           0.087514
6           0.264771
7           0.568004
8           0.714604
9           0.146600
10          0.320513
11          0.097547
12          0.738573
13          0.104794
14          0.696767
15          0.073579
16          0.522854
17          0.076923
18          0.059643
19          0.463768
20          0.198997
21          0.101449
22          0.338350
23          0.236901
24          0.627090
25          0.171126
26          0.031773
27          0.261427
28          0.171126
29          0.714604
..               ...
850         0.076923
851         0.153289
852         0.916388
853         0.222965
854         0.380156
855         0.264771
856         0.104794
857         0.244147
858         0.219621
859         0.233556
860         0.111483
861         0.087514
862         0.087514
863         0.122074
864         0.132664
865         0.226867
866         0.188406
867         0.069677
868         0.459866
869         0.171126
870         0.076923
871         0.926979
872         0.498328
873         0.115385
874         0.366221
875         0.512263
876         0.268673
877         0.617057
878         0.250836
879         0.606466

[880 rows x 1 columns]

In [67]:
temperature_l12w_test_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TESTING_May-4_August-03_2018.csv', sep=',', header=True, index=False)